File: examples/typescript/05-custom-sandboxed-tool/src/main.ts
/**
 * spore-core example 05 — a custom tool you write yourself.
 *
 * Examples [`03`](../../03-tool-use) and [`04`](../../04-filesystem-agent)
 * showed the two *built-in* tool paths: hand-rolling the harness-loop
 * `ToolRegistry` (03) and registering the shipped catalogue with
 * `.tools(StandardTools.codingSet())` (04). This example shows the third and
 * most important path — **bringing your own tool** — and the public extension
 * point that makes it possible: the `Tool` interface.
 *
 * ## The two custom tools
 *
 * Both are plain classes implementing `Tool` (see `src/tools/`):
 *
 * - **`remember(key, value)`** — persists a fact into the run store. It MUTATES
 *   shared state, so it is not `read_only`.
 * - **`recall(key)`** — reads a fact back out. It only reads, so it is
 *   `read_only` + `idempotent`.
 *
 * ## The seam each tool uses
 *
 * Each tool's `execute(call, sandbox, ctx)` receives two seams: a
 * `SandboxProvider` (the environment — unused here, these tools never touch the
 * filesystem) and a `ToolContext` (the storage seam: `ctx.runStore` +
 * `ctx.sessionId`). Facts are keyed under `fact:{key}` so they cannot collide
 * with reserved catalogue keys.
 *
 * ## The pattern: implement `Tool` → wrap in a `StandardTool` → `.tool()`
 *
 * 1. Implement `Tool` (a class with `name` + `execute`).
 * 2. Bundle the impl with its schema as a `StandardTool` (`{ implementation,
 *    schema }`) so the two can never drift.
 * 3. Register each with `.tool(...)`. The harness wires the sandbox and a
 *    per-run `ToolContext` automatically — **the harness doesn't change, only
 *    what you register does.**
 *
 * Two builder differences from 04: there is no `.tools(...)` catalogue, and no
 * explicit `.sandbox(...)` / `.storage(...)`. `build()` defaults storage to an
 * in-memory provider whenever `.tool()` tools are present, so the run store
 * works for free.
 *
 * ## Run it
 *
 * ```sh
 * ollama serve &
 * ollama pull llama3.2
 * pnpm install
 * pnpm start
 * ```
 */

import {
  HarnessBuilder,
  OLLAMA_DEFAULT_BASE_URL,
  OllamaModelInterface,
  SessionId,
  newTask,
  type HarnessStreamEvent,
  type toolRegistry,
} from "@spore/core";

import { RecallTool } from "./tools/recall.js";
import { RememberTool } from "./tools/remember.js";

type StandardTool = toolRegistry.StandardTool;

const SYSTEM_PROMPT =
  "You are a research agent with a memory. Research the topic the user gives " +
  "you across several turns. As you discover each fact, call `remember` to " +
  "store it under a short, stable key (e.g. 'habitat', 'diet'). Keep track of " +
  "the keys you use. When you have gathered enough facts, call `recall` on each " +
  "key you remembered, then write a final summary built ONLY from the recalled " +
  "facts. Act using tools — do not just describe.";

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const modelId =
    argValue(args, "--model") ?? process.env.SPORE_OLLAMA_MODEL ?? "llama3.2";
  const baseUrl = process.env.SPORE_OLLAMA_BASE_URL ?? OLLAMA_DEFAULT_BASE_URL;

  const prompt =
    argValue(args, "--prompt") ??
    "Research the common octopus. Remember a few key facts (habitat, diet, " +
      "lifespan, intelligence), then recall them and write a short summary.";

  // Same `conversational` harness as 03 / 04 — the substantive change is that we
  // register two tools WE wrote (`.tool(...)`) instead of a catalogue preset. No
  // `.sandbox(...)` (these tools ignore it) and no `.storage(...)` (build()
  // defaults to in-memory storage when `.tool()` tools are present).
  const model = OllamaModelInterface.withBaseUrl(modelId, baseUrl);
  const remember: StandardTool = {
    implementation: new RememberTool(),
    schema: RememberTool.schema(),
  };
  const recall: StandardTool = {
    implementation: new RecallTool(),
    schema: RecallTool.schema(),
  };
  const harness = HarnessBuilder.conversational(model)
    .tool(remember)
    .tool(recall)
    .systemPrompt(SYSTEM_PROMPT)
    .build();

  const task = newTask(prompt, SessionId.generate(), {
    kind: "re_act",
    max_iterations: 12,
  });

  console.log(`model  : ${modelId}`);
  console.log(`tools  : remember, recall`);
  console.log(`prompt : ${prompt}\n`);

  // Print each turn (Think) and each tool call + result (Act / Observe) from
  // harness STREAM events — the builder dispatches our tools internally, just as
  // it does the catalogue in 04.
  const onStream = (event: HarnessStreamEvent): void => {
    switch (event.kind) {
      case "turn_start":
        console.log(`think  · turn ${event.turn}`);
        break;
      case "tool_call":
        console.log(
          `    act    → ${event.name}(${JSON.stringify(event.args ?? {})})`,
        );
        break;
      case "tool_result": {
        const tag = event.is_error ? "obs(err)" : "obs ";
        console.log(`    ${tag}→ ${truncate(event.content ?? "", 200)}`);
        break;
      }
      default:
        break;
    }
  };

  let result;
  try {
    result = await harness.run({ task, on_stream: onStream });
  } catch (err) {
    console.error(
      `\ncould not reach the model — is Ollama running at ${baseUrl}? (\`ollama serve\`)\n${err instanceof Error ? err.message : String(err)}`,
    );
    process.exit(1);
  }

  if (result.kind === "success") {
    console.log(`\nsummary (${result.turns} turn(s)): ${result.output}`);
    return;
  }
  console.error(`\nrun did not succeed: ${JSON.stringify(result)}`);
  process.exit(1);
}

function argValue(args: string[], flag: string): string | undefined {
  const i = args.indexOf(flag);
  return i >= 0 ? args[i + 1] : undefined;
}

/** Keep observe lines readable — recalled facts can be long. */
function truncate(s: string, max: number): string {
  const flat = s.replace(/\n/g, " ");
  const chars = [...flat];
  return chars.length <= max ? flat : `${chars.slice(0, max).join("")}…`;
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: examples/rust/05-custom-sandboxed-tool/src/main.rs
//! spore-core example 05 — a custom tool you write yourself.
//!
//! Examples [`03`](../../03-tool-use) and [`04`](../../04-filesystem-agent)
//! showed the two *built-in* tool paths: hand-rolling the harness-loop
//! `ToolRegistry` (03) and registering the shipped catalogue with
//! `.tools(StandardTools::coding_set())` (04). This example shows the third and
//! most important path — **bringing your own tool** — and the public extension
//! point that makes it possible: the [`Tool`] trait.
//!
//! ## The two custom tools
//!
//! Both are defined in [`mod tools`](tools) as plain `impl Tool`:
//!
//! - **`remember(key, value)`** — persists a fact into the run store
//!   ([`tools::remember`]). It MUTATES shared state, so it is not `read_only`.
//! - **`recall(key)`** — reads a fact back out ([`tools::recall`]). It only
//!   reads, so it is `read_only` + `idempotent`.
//!
//! ## The trait used, and the seam it exposes
//!
//! Each tool implements [`spore_core::Tool`]. The `execute` method receives two
//! seams: a `SandboxProvider` (the environment — unused here, these tools never
//! touch the filesystem) and a [`ToolContext`](spore_core::ToolContext) (the
//! storage seam: `run_store()` + `session_id()`). Facts are keyed under
//! `fact:{key}` so they cannot collide with reserved catalogue keys.
//!
//! ## The pattern: `impl Tool` → `StandardTool::new` → `.tool()`
//!
//! 1. Implement [`Tool`](spore_core::Tool).
//! 2. Bundle the impl with its schema via
//!    [`StandardTool::new`](spore_core::StandardTool) so the two can never drift.
//! 3. Register each with `.tool(...)`. The harness wires the sandbox and a
//!    per-run `ToolContext` automatically — **the harness doesn't change, only
//!    what you register does.**
//!
//! Two builder differences from 04: there is no `.tools(...)` catalogue, and no
//! explicit `.sandbox(...)` / `.storage(...)`. `build()` defaults storage to an
//! in-memory provider whenever `.tool()` tools are present, so the run store
//! works for free.
//!
//! (No `// SPEC QUESTION:` markers remain — all design points are resolved.)
//!
//! ## Run it
//!
//! ```sh
//! ollama serve &
//! ollama pull llama3.2
//! cargo run
//! ```

mod tools;

use spore_core::{
    Harness, HarnessBuilder, HarnessRunOptions, HarnessStreamEvent, LoopStrategy,
    OllamaModelInterface, RunResult, SessionId, StandardTool, Task,
};

use tools::recall::RecallTool;
use tools::remember::RememberTool;

const SYSTEM_PROMPT: &str = "You are a research agent with a memory. Research the topic the user \
     gives you across several turns. As you discover each fact, call `remember` to store it under \
     a short, stable key (e.g. 'habitat', 'diet'). Keep track of the keys you use. When you have \
     gathered enough facts, call `recall` on each key you remembered, then write a final summary \
     built ONLY from the recalled facts. Act using tools — do not just describe.";

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = std::env::args().skip(1).collect();
    let model_id = arg_value(&args, "--model")
        .or_else(|| std::env::var("SPORE_OLLAMA_MODEL").ok())
        .unwrap_or_else(|| "llama3.2".to_string());
    let base_url = std::env::var("SPORE_OLLAMA_BASE_URL")
        .unwrap_or_else(|_| OllamaModelInterface::DEFAULT_BASE_URL.to_string());

    let prompt = arg_value(&args, "--prompt").unwrap_or_else(|| {
        "Research the common octopus. Remember a few key facts (habitat, diet, lifespan, \
         intelligence), then recall them and write a short summary."
            .to_string()
    });

    // Same `conversational` harness as 03 / 04 — the substantive change is that
    // we register two tools WE wrote (`.tool(...)`) instead of a catalogue
    // preset. No `.sandbox(...)` (these tools ignore it) and no `.storage(...)`
    // (build() defaults to in-memory storage when `.tool()` tools are present).
    let model = OllamaModelInterface::with_base_url(&model_id, base_url);
    let harness = HarnessBuilder::conversational(model)
        .tool(StandardTool::new(
            Box::new(RememberTool::new()),
            RememberTool::schema(),
        ))
        .tool(StandardTool::new(
            Box::new(RecallTool::new()),
            RecallTool::schema(),
        ))
        .system_prompt(SYSTEM_PROMPT)
        .build();

    let task = Task::new(
        prompt.clone(),
        SessionId::generate(),
        LoopStrategy::ReAct { max_iterations: 12 },
    );
    // Print each turn (Think) and each tool call + result (Act / Observe) from
    // harness STREAM events — the builder dispatches our tools internally, just
    // as it does the catalogue in 04.
    let options = HarnessRunOptions::new(task).with_stream(Box::new(
        |event: HarnessStreamEvent| match event {
            HarnessStreamEvent::TurnStart { turn } => println!("think  · turn {turn}"),
            HarnessStreamEvent::ToolCall { name, args, .. } => {
                println!("    act    → {name}({args})");
            }
            HarnessStreamEvent::ToolResult {
                is_error, content, ..
            } => {
                let tag = if is_error { "obs(err)" } else { "obs " };
                println!("    {tag}→ {}", truncate(&content, 200));
            }
            _ => {}
        },
    ));

    println!("model  : {model_id}");
    println!("tools  : remember, recall");
    println!("prompt : {prompt}\n");
    match harness.run(options).await {
        RunResult::Success { output, turns, .. } => {
            println!("\nsummary ({turns} turn(s)): {output}");
            Ok(())
        }
        other => {
            eprintln!("\nrun did not succeed: {other:?}");
            std::process::exit(1);
        }
    }
}

fn arg_value(args: &[String], flag: &str) -> Option<String> {
    args.iter()
        .position(|a| a == flag)
        .and_then(|i| args.get(i + 1).cloned())
}

/// Keep observe lines readable — recalled facts can be long.
fn truncate(s: &str, max: usize) -> String {
    let s = s.replace('\n', " ");
    if s.chars().count() <= max {
        s
    } else {
        let cut: String = s.chars().take(max).collect();
        format!("{cut}…")
    }
}
